backend/remote/backend_context.go

```diff
@@ -125,8 +125,39 @@ func (b *Remote) Context(op *backend.Operation) (*terraform.Context, statemgr.Fu
 		}
 
 		if op.Variables != nil {
-			variables, varDiags := backend.ParseVariableValues(op.Variables, config.Module.Variables)
+			variables, varDiags := backend.ParseVariableValuesWithoutDefaults(op.Variables, config.Module.Variables)
 			diags = diags.Append(varDiags)
+			for name, vc := range config.Module.Variables {
+				if _, defined := variables[name]; defined {
+					continue
+				}
+
+				if vc.Required() {
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "No value for required variable",
+						Detail:   fmt.Sprintf("The root module input variable %q is not set, and has no default value. Set it in your remote workspace:\nhttps://%s/app/%s/%s/variables", name, b.hostname, b.organization, op.Workspace),
```

Comment:
I have a question: Is it theoretically possible for this crafted link to be incorrect? For example (in theory), could a user have their own implementation of terraform cloud that works with the remote backend but uses a different url ("https://my-example.com/weirdsubthing/app/...")?

Reply:
I wondered about this too but we're making this same assumption in a few other places, so it felt like a safe pattern to continue.

Example pointing to this url if a user provides a run-specific variable: https://github.com/hashicorp/terraform/blob/c1f69fba031182af7d89b23524d77e87dfc5cba4/backend/remote/backend_apply.go#L81